hibernate-core/src/main/java/org/hibernate/event/internal/AbstractSaveEventListener.java

```diff
@@ -248,7 +244,7 @@ protected Serializable performSaveOrReplicate(
 
 		Serializable id = key == null ? null : key.getIdentifier();
 
-		boolean shouldDelayIdentityInserts = shouldDelayIdentityInserts( requiresImmediateIdAccess, source, persister );
+		boolean shouldDelayIdentityInserts = !isPartOfTransaction( source ) && !requiresImmediateIdAccess;
```

Comment:
Hibernate may not do the right thing when:  `!isPartOfTransaction( source ) && requiresImmediateIdAccess`.

IIUC,  `shouldDelayIdentityInserts` will be set to `false`, and Hibernate will perform the insert (using auto-commit?), even if `hibernate.allow_update_outside_transaction=false` (the default). Shouldn't `TransactionRequiredException` be thrown?

I'll address `#isPartOfTransaction` below.

Background (as I understand it):

IIUC, Hibernate requires that the entity have an ID when `Session#save` (which returns the ID) or `Session#saveOrUpdate` is used to persist a transient entity, so `requiresImmediateIdAccess == true` in those cases.

`Session#persist` does not require that the entity have an ID until after flush so `requiresImmediateIdAccess == false` in that case.

Reply:
I suppose this could be handled in a separate jira issue.